### PR TITLE
Create tarballs for packages and upload as artifact for PRs

### PR DESCRIPTION
**Pull request by johnpmitsch**
This is a way to test out package changes in PRs without uploading anything to npm. More information below

The best way to test this is to use a PR that upgrades jest on my fork https://github.com/johnpmitsch/foreman-js/pull/14#issuecomment-697498083 - The action is merged into master there so we can test out what a PR upgrading a package would look like with these actions

### CODE REVIEW

**Comment by johnpmitsch** [comment]
@sharvit @jeremylenz This is one potential solution I have for testing package builds of PRs. 

You can create a tarball with 'npm pack', and that can be used to npm install, [see 'tarball path' and 'tarball url' options here](https://docs.npmjs.com/cli-commands/install.html)

Github actions supports uploading a local artifact, which is made [available to download in the action](https://github.com/actions/upload-artifact#where-does-the-upload-go)

So we can use an action create a tarball and upload to the github action as an artifact, then use that artifact url from github, updating foreman's package.json to point to the tarball url. An advantage to this approach is the npm package resolution would happen naturally, rather than manually modifying `node_modules`, which sometimes doesn't resolve as expected.

lerna doesn't support a 'pack' command for every package, there is [an issue requesting it](https://github.com/lerna/lerna/issues/2134), though it should be easy to loop through `packages/*`, pack each one, and upload the tarball artifact.

This would bypass the need to use npm at all and not expose any secrets or leave us vulnerable the action being exploited.

I am just testing out on the `test` package, but not sure how to run a new action from a PR :thinking:

**Comment by johnpmitsch** [comment]
Here is an example locally:

1. I upgrade `jest` in the `test` package:
```diff
diff --git a/packages/test/package.json b/packages/test/package.json     
index b00db8c..4bdfe27 100644                                            
--- a/packages/test/package.json                                         
+++ b/packages/test/package.json                                         
@@ -29,7 +29,7 @@                                                        
     "@theforeman/builder": "^4.15.1",
     "@theforeman/vendor-core": "^4.15.1",
     "axios-mock-adapter": "^1.1.7",
-    "babel-jest": "^24.9.0",                                            
+    "babel-jest": "^26.3.0",                                            
     "babel-plugin-dynamic-import-node": "^2.0.0",
     "cheerio": "0.22.0",
     "commander": "^4.0.1",   
@@ -38,7 +38,7 @@                                                        
     "enzyme-adapter-react-16": "^1.15.2",
     "enzyme-to-json": "^3.4.3",
     "identity-obj-proxy": "^3.0.0",
-    "jest-cli": "^24.9.0",                                              
+    "jest-cli": "^26.4.0",                                              
     "jest-prop-type-error": "^1.1.0",
     "raf": "^3.4.0",      
     "react-redux-test-utils": "^0.2.0"
```

2. Run `npm pack` in `packages/test` to create a tarball
```
[vagrant@centos7-katello-devel-stable test{jest_upgrade}]$ npm pack
...
theforeman-test-4.15.1.tgz
```

3. Have foreman point at the local tarball
```
[vagrant@centos7-katello-devel-stable foreman{develop}]$ cat package.json  | grep @theforeman/test
    "@theforeman/test": "/home/vagrant/foreman-js/packages/test/theforeman-test-4.15.1.tgz",
```
4. `rm -rf node_modules && rm package-lock.json` in `foreman`. This is needed because the version of `@theforeman/test` has not changed and even if it did, it still meets the specifications, so we need a fresh install.
5. `npm i`
6. jest has been updated and npm is resolving to the updated one:
```
[vagrant@centos7-katello-devel-stable foreman{develop}]$ npm ls jest-cli
TheForemanDevDeps@2.3.0 /home/vagrant/foreman
└─┬ @theforeman/test@4.15.1
  └── jest-cli@26.4.2 

[vagrant@centos7-katello-devel-stable foreman{develop}]$ npx jest --version
26.4.2
```

In the github actions version of this, github actions will do the `npm pack` (no step 2) and in step 3, foreman will point at the artifact URL.

**Comment by johnpmitsch** [comment]
On a side note: assuming there is no caveats that I am missing, I much prefer this method of testing locally. Without any symlinks, no files are getting overwritten in `foreman-js` when I `npm i` in Foreman (which I always seem to do absentmindedly)

**Comment by johnpmitsch** [comment]
This still needs docs, but I put an example of how this would function here https://github.com/johnpmitsch/foreman-js/pull/14 on my fork, where the action is already merged.

Could someone test that PR change out and confirm the workflow works as expected? Feel free to open up other changes against my fork to test things out.

**Comment by johnpmitsch** [comment]
We may be able to use https://github.com/dorny/paths-filter to only build the artifact if the `packages/package_name` subdirectory changed :thinking:

This way only relevant packages would be built as artifacts, though i'm not sure how that would fit in with the matrix, the jobs may all still need to run but would abort if there were not changes.

They are very quick jobs so not sure if it's a concern!

**Comment by johnpmitsch** [comment]
Updated with docs

**Comment by johnpmitsch** [comment]
I updated to check the paths

> We may be able to use https://github.com/dorny/paths-filter to only build the artifact if the `packages/package_name` subdirectory changed

I updated to check the path for changes and only run the steps if the path for the package has changed. The job still runs and shows up in actions, but will not perform any further steps if there are no changes. Even though the jobs are quite fast, this will prevent unnecessary work and there will only be relevant artifacts created.

I had to use an `if` statement in each step, there is no way to do branched conditionals inside of a job. Normally you would use another job and depend on the previous one,  but we wouldn't have the context of the matrix. Maybe there is another way, but I didn't have luck trying out other methods. I'm personally ok with a little repetition

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -85,6 +85,35 @@ git push origin my-branch
 
 5. Open [this project on GitHub](https://github.com/theforeman/foreman-js), then click “Compare & pull request”.
 
+## Testing Pull Requests
+
+1. Open pull request in Github and go to the tasks tab.
```
Comment (jeremylenz) [inline CONTRIBUTING.md:90]:
```suggestion
1. Open the pull request in Github and go to the __Checks__ tab.
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -85,6 +85,35 @@ git push origin my-branch
...
+2. Expand `Build tarballs for testing` jobs and open the relevant action. For example, the job for the test package would be `build-tarball (test)`.
```
Comment (jeremylenz) [inline CONTRIBUTING.md:91]: The artifact seems to be available no matter which action I select in the nav bar.
```suggestion
2. Expand `Build tarballs for testing` and open any of the associated actions. For example, the job for the test package would be `build-tarball (test)`.
```

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -85,6 +85,35 @@ git push origin my-branch
...
+cd /path/to/download_dir && unzip foreman-js-test-pr-123.tgz.zip
+scp theforeman-test-4.15.1.tgz vagrant@centos7-katello-devel-stable.example.com:~
+```
+  - Side note: ideally we could use the tarball URL directly but looks like there is a permission issue that will make it hard to use with npm: https://github.com/actions/upload-artifact/issues/51
+5. Update package.json in foreman or plugin to use local tarball:
+```
+$ cat package.json | grep theforeman-test
+    "@theforeman/test": "/home/vagrant/theforeman-test-4.15.1.tgz",
+```
+6. `rm -rf node_modules && rm package-lock.json` in foreman or plugin.
```
Comment (jeremylenz) [inline CONTRIBUTING.md:105]: This makes it about as much work as the other `npm i; npm run build; npm run link` workflow.  I wonder if there's any way to tell npm to be more selective or just to replace a single package..

File: .github/workflows/build-tarballs.yml [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+      with:
+        node-version: 12.x
+    - name: Checkout foreman-js
+      if: ${{ steps.changes.outputs.matched == 'true' }}
+      uses: actions/checkout@v2
+    - name: Package tarball
+      if: ${{ steps.changes.outputs.matched == 'true' }}
+      run:  npm pack 
+      working-directory: ${{ github.workspace }}/packages/${{ matrix.package }}
+    - name: Upload tarball locally
```
Comment (jeremylenz) [inline .github/workflows/build-tarballs.yml:51]: Would be awesome if this action could comment on the PR with the direct download link to the artifact.  (Or at least tell you which packages had changes)

File: CONTRIBUTING.md [hunk 5]
```diff
@@ -85,6 +85,35 @@ git push origin my-branch
...
+1. Open pull request in Github and go to the tasks tab.
+2. Expand `Build tarballs for testing` jobs and open the relevant action. For example, the job for the test package would be `build-tarball (test)`.
+3. Expand the artifacts dropdown on the top right and download the appropiate artifact. For example, the test package will be titled `foreman-js-test-pr-###.tgz`.
+4. Unzip and copy the unpacked tarball to your development box or the appropiate location:
+```bash
+# download `foreman-js-test-pr-123.tgz.zip` from `artifacts` dropdown on top right of github action
+cd /path/to/download_dir && unzip foreman-js-test-pr-123.tgz.zip
+scp theforeman-test-4.15.1.tgz vagrant@centos7-katello-devel-stable.example.com:~
+```
+  - Side note: ideally we could use the tarball URL directly but looks like there is a permission issue that will make it hard to use with npm: https://github.com/actions/upload-artifact/issues/51
```
Comment (jeremylenz) [inline CONTRIBUTING.md:99]: Would it work to have a shell script download the tarball with `curl`?
If so, it wouldn't be too hard to have a script that automates steps 4 through 8-- you could just pass in the url.

File: .github/workflows/build-tarballs.yml (hunk 4 above)
Comment (jeremylenz) [reply .github/workflows/build-tarballs.yml:51]: this may be prevented because of the issue with secrets and PRs from forks.  :confused:

**Review by jeremylenz — COMMENTED** [review]
Nice work @johnpmitsch !!

Some initial comments below.. I didn't actually download the artifact and test but I did look at this and the fork PR.

Personally I may not switch completely to this new method (just because fresh `npm install`s are so slow) but I totally support having it as an option!  Also it may get around some issues we're having such as https://github.com/theforeman/foreman-js/issues/213

File: CONTRIBUTING.md (hunk 3 above)
Comment (johnpmitsch) [reply CONTRIBUTING.md:105]: The benefits to me are:
A) Module resolution happens naturally, so any executables provided by a package function as expected and interactions with other packages outside of @theforeman ones resolve as expected.
B) No longer having to worry about symlinked files being modified. I've too often symlinked to foreman, forgot about it, and then ran `npm i` and overwrote my foreman-js folder.
C) It's easier for those outside of those not knowledgeable about this repo to review PRs

I get that it takes some time to reinstall everything, but it's unattended and we can be sure that the packages are resolving as they will once the new @theforeman package is published. To me it's not about saving time, more having a way to test that will function the same way as the published package will.

Of course I am fine with keeping `npm run link`, I can understand if people are used to that workflow or prefer it.

File: CONTRIBUTING.md (hunk 5 above)
Comment (johnpmitsch) [reply CONTRIBUTING.md:99]: From the issue, it sounds like `curl -v -H "Authorization: token $GITHUB_TOKEN" $TARBALL_URL` would work

We could try out the curl command and give it as an option if it works as expected. I would just want to make sure the permissions don't block it for non-maintainers (or we make that distinction clear in the docs)

File: .github/workflows/build-tarballs.yml (hunk 4 above)
Comment (johnpmitsch) [reply .github/workflows/build-tarballs.yml:51]: A comment with the download link and some info on how to set it up would be really nice, I can look into it!

**Comment by johnpmitsch** [comment]
@jeremylenz Updated! I modified the action a bit to remove all the file paths from the matrix, we can just use `packages/$PACKAGE/**` for the path check.

I also provided instructions on how to download the artifact directly in the docs, it isn't super straightforward, but once you get the hang of it, it's not bad. Unfortunately I can't see an easy way to get an artifact ID to use in a comment. Github has made it really hard to get an artifact by URL, you have to go through the API and deduce the the artifact ID from the url in the action. (there are many issues complaining about this) There could probably be a scripted step using the API or a custom action to do this, but not sure if I will have the time to look into that for this PR!

**Review by jeremylenz — APPROVED** [review]

**Pull request merged**